index.bs

```diff
@@ -2602,8 +2655,247 @@ Note: Future variations of this specification will retain the invariant that
+place of a realm, in which case the realm used is the realm of the browsing
+context's active document.
+
+The method returns the value of executing the provided script, unless it returns
+a promise and <code>awaitPromise</code> is true (which is the default), in which
+case the resolved value of the promise is returned.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      ScriptEvaluateCommand = {
+        method: "script.evaluate",
+        params: ScriptEvaluateParameters
+      }
+
+      ScriptEvaluateParameters = {
+        expression: text,
+        ?awaitPromise: bool,
+        target: RealmOrBrowsingContext,
+      }
+      </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+      ScriptEvaluateResult = {
+         result: RemoteValue,
+         exceptionDetails: Exception
+      }
+
+      ScriptGetRealmsResult = {
+        realms: RealmInfoList
+      }
+    </pre>
+   </dd>
+</dl>
+
+TODO: Add timeout argument. It's not totally clear how this ought to work; in
+Chrome it seems like the timeout doesn't apply to the promise resolve step, but
+that likely isn't what clients want.
+
+The [=remote end steps=] with |command parameters| are:
+
+1. Let |realm| be the result of [=trying=] to [=get a realm from a target=]
+   given the value of the <code>target</code> field of |command parameters|.
+
+1. Let |environment settings| be the [=environment settings object=] whose
+   [=realm execution context=]'s Realm component is |realm|.
+
+1. Let |source| be the value of the <code>expression</code> field of |command
+   parameters|.
+
+1. Let |await promise| be the value of the <code>awaitPromise</code> field of
+   |command parameters|, if present, or <code>true</code> otherwise.
+
+1. Let |options| be the [=default classic script fetch options=].
+
+1. Let |base URL| be the [=API base URL=] of |environment settings|.
+
+1. Let |script| be the result of [=create a classic script=] with |source|,
+   |environment settings|, |base URL|, and |options|.
+
+1. [=Prepare to run script=] with |environment settings|.
+
+1. Set |evaluation status| to [=ScriptEvaluation=](|script|'s record).
+
+1. If |evaluation status|.\[[Type]] is <code>normal</code>, |await promise| is
+   true, and [=IsPromise=](|evaluation status|.\[[Value]]):
+
+   1. Set |evaluation status| to <a spec=ECMASCRIPT>Await</a>(|evaluation status|.\[[Value]]).
+
+1. [=Clean up after running script=] with |environment settings|.
+
+1. Let |value| be |evaluation status|.\[[Value]]
+
+1. If |evaluation status|.\[[Type]] is <code>throw</code>, let |exception
+   details| be the result of [=get exception details=] given |evaluation
+   status|. Otherwise let |exception details| be null.
+
+1. Let |result| be the result of [=serialize as a remote value=] given |value|.
+
+1. Return a new map matching the <code>ScriptEvaluateResult</code> production,
```

Comment:
Is it possible for a script evaluation to both throw and return a result? Otherwise, I think ScriptEvaluateResult would be better expressed as a sum type which can be either { result } or { exceptionDetails }.

Reply:
I thought so too, but that's at least not quite how CDP works. It has both a `result` field with the object that was thrown and also a `exceptionDetails` field with details of the error. For example `throw 2` ends up as:
```
{
    "result": {
        "type": "number",
        "value": 2,
        "description": "2"
    },
    "exceptionDetails": {
        "exceptionId": 11,
        "text": "Uncaught",
        "lineNumber": 0,
        "columnNumber": 0,
        "scriptId": "78",
        "stackTrace": {
            "callFrames": [
                {
                    "functionName": "",
                    "scriptId": "78",
                    "url": "",
                    "lineNumber": 0,
                    "columnNumber": 0
                }
            ]
        },
        "exception": {
            "type": "number",
            "value": 2,
            "description": "2"
        }
    }
}
```

With a non-primitive value things are a little more interesting, `throw {}` gives:
```
{
    "result": {
        "type": "object",
        "className": "Object",
        "description": "Object",
        "objectId": "5263362622729908126.1.21",
        "preview": {
            "type": "object",
            "description": "Object",
            "overflow": false,
            "properties": []
        }
    },
    "exceptionDetails": {
        "exceptionId": 18,
        "text": "Uncaught",
        "lineNumber": 0,
        "columnNumber": 0,
        "scriptId": "91",
        "stackTrace": {
            "callFrames": [
                {
                    "functionName": "",
                    "scriptId": "91",
                    "url": "",
                    "lineNumber": 0,
                    "columnNumber": 0
                }
            ]
        },
        "exception": {
            "type": "object",
            "className": "Object",
            "description": "Object",
            "objectId": "5263362622729908126.1.22",
            "preview": {
                "type": "object",
                "description": "Object",
                "overflow": false,
                "properties": []
            }
        }
    }
}
```
`result` and `exceptionDetails.exception` refer to the same object but the `objectID` is different. I don't know if CDP is packing some information into the id so it's possible to tell it's the same object, rather than using it as an opaque identifier.

Anyway, I haven't found a case where `result` and `exceptionDetails.exception` can be different, so I think we should do one or the other rather than both. We could do that as "either `result` or `exceptionDetails`" or we could always value `result` and just use `exceptionDetails` for additional information.

Reply:
FWIW
There are [3 types of IDs in CDP](https://github.com/WICG/devtools-protocol/issues/9#issuecomment-367831808):
1. `RemoteObjectId`
1. `NodeId`
1. `BackendNodeId`

According to the [`bindObject` logic](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/inspector/injected-script.cc;drc=9481da96106e5863fdd22bf54a0a6c264e6c56f4;l=1003), all the objects in the same context will have the same first 2 numbers in the ID, and the incremental last number even for the same object.